### ISSUE TITLE
Tests: `expect`:  It is not `assert`.  A failed `expect` should not stop execution

### ISSUE THREAD

**Issue by WorikQCI** [Dart-Code/Dart-Code#4470]
**Is your feature request related to a problem? Please describe.**
When I build a unit test to test a function I often want to test a set on inputs.

Some will fail, some will succeed.

I am testing a function so each call to it is independant

So if an `expect` fails I wish the `expect` calls after it to still be executed.  Otherwise I am wasting time.

As it stands the `expect` call is entirely superfluous as there is already an `assert` that does basically the same thing.

```
  test("except==assert", () {
    print("true, true");
    expect(true, true);
    print("false, false");
    expect(false, false);
    print("false, true");
    expect(false, true);
    print("false, false");
    expect(false, false);
    print("true, false");
    expect(true, false);
  });
```

Outputs:
```
00:00 ESC[32m+0ESC[0mESC[33m ~1ESC[0m: except==assertESC[0m

true, true

false, false

false, true

00:00 ESC[32m+0ESC[0mESC[33m ~1ESC[0mESC[31m -1ESC[0m: except==assert ESC[1mESC[31m[E]ESC[0mESC[0m

  Expected: <true>
    Actual: <false>
  

  package:test_api/src/expect/expect.dart 134:31     fail
  package:test_api/src/expect/expect.dart 129:3      _expect
  package:test_api/src/expect/expect.dart 46:3       expect
  test/issue_demo_test.dart 27:5                     main.<fn>
  package:test_api/src/backend/declarer.dart 215:19  Declarer.test.<fn>.<fn>
  ===== asynchronous gap ===========================
  package:test_api/src/backend/declarer.dart 213:7   Declarer.test.<fn>
  ===== asynchronous gap ===========================
  package:test_api/src/backend/invoker.dart 257:7    Invoker._waitForOutstandingCallbacks.<fn>
  

00:00 ESC[32m+0ESC[0mESC[33m ~1ESC[0mESC[31m -1ESC[0m: ESC[31mSome tests failed.ESC[0m

```

It detected the first: `    expect(false, true);` but failed to detect the following: `expect(true, false);`

**Comment by DanTup** [comment]
I don't think there's anything the VS Code extension can do here. The `expect` method calls `fail()` when the expectation is not met, which `throw`s:

https://github.com/dart-lang/matcher/blob/61f43473951ce425982c2cfb694f3fbe50b1655e/lib/src/expect/expect.dart#L149

This prevents any more test code from running.

@jakemac53 I can see an argument for both ways here... on one hand, you might want to see multiple failed expectations, but on the other hand most tests are written such that after the first expectation fails, the rest of the test is probably invalid.

Is this something that's come up before? I thought it would have, but I couldn't find any related issues in the `package:test` repo.

**Comment by jakemac53** [comment]
It has come up in the past for sure, but I also am not sure exactly where those issues or discussion might be. I think the current behavior is the best _default_ behavior though.

> As it stands the `expect` call is entirely superfluous as there is already an `assert` that does basically the same thing.

These are _not_ the same thing really. The main difference is that `expect` supports matchers, as well as implicitly applying the `equals` matcher. Ultimately this gives you a lot better output than a simple `assert` does.

## Side note

We are currently working on making it so that custom frameworks can more easily be written to work hand in hand with package:test. The main reason is we want to introduce the `checks` package, which doesn't use `expect` at all.

To that end, `expect`, and essentially everything except for `test`, `group`, and the annotations, are eventually all going to go away and no longer be a part of the `package:test/test.dart` import.

This is all leading us to expose some better hooks (from test_api) for authors that want to make their own expectation/matcher frameworks.

So basically what I am saying, is in the near future you will be able to make your own library which behaves how you want :). And it probably won't really be all that much work, either.

**Comment by WorikQCI** [comment]
This below sounds really bad.

What I, as a programmer need, is a simple `expect` function.  All the extra services that are provided by `matcher` are easy enough to do myself.

I do not need easy things made easier.  

The way expect works now is I have to break up test functions, and repeat code, in arbitrary ways.  That is not easy.

So having `expect`, when it fails, allow execution to flow through would make some hard things easier.  That is Good Work.

Pretty much all testing comes down to: Create state.  Check the state.  

All the pieces are there, except that blasted `expect` function is an  `assert`, essentially.

You do have serious problems with the output of the test framework.  It is hard to parse and unintuitive - that is another issue for another day, but from my perspective as a programmer using your software (thank you) that is where I would like to see real improvements.

None of this below is necessary.  Somebody is thinking too hard!

```
To that end, expect, and essentially everything except for test, group, and the annotations, are eventually all going to go away and no longer be a part of the package:test/test.dart import.

This is all leading us to expose some better hooks (from test_api) for authors that want to make their own expectation/matcher frameworks (package:checks will not have any special power).


```

**Comment by DanTup** [comment]
@jakemac53 great, thanks for the info!

@WorikQCI as far as I can tell, what @jakemac53 is describing will allow you to do exactly what you want - you can change `expect` to not throw, and to format failures however you'd like. I suspect once this work is done, there will be more choices of packages that provide these kinds of behaviours for you.

Closing this as I don't think there's anything to do in the VS Code extension, I think this will all come from `package:test` changes. @jakemac53 if there's an issue that can be followed for progress of that new work, I'd be interested in seeing the results (and a link here could be useful for others discovering this issue).

Thanks!

**Comment by jakemac53** [comment]
I don't know of any specific tracking issue (we don't really use the issue tracker for feature development). cc @natebosch

**Comment by natebosch** [comment]
You can add 👍 , 👎 , or feedback about a non-throwing way to collect mismatches and report them all together on https://github.com/dart-lang/test/pull/1898

**Comment by WorikQCI** [comment]
This is not anything to do with VS Code.

Why do you keep brining up "VS Code Extensions"

I see the same thing using the command line.  I would rather use the command line.

**Comment by jakemac53** [comment]
> This is not anything to do with VS Code.

This issue tracker is for the VsCode extension 🤣 .

**Comment by DanTup** [comment]
lol, np 😄

It's not that uncommon for upstream Dart issues to be filed here anyway (it's not always obvious what functionality comes directly from the VS Code extension and what's coming from Dart/Flutter SDK tools the extension calls).
